Mathlib/Analysis/Convex/Radon.lean

```diff
@@ -0,0 +1,232 @@
+/-
+Copyright (c) 2023 Vasily Nesterov. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Vasily Nesterov
+-/
+import Mathlib.Analysis.Convex.Combination
+import Mathlib.LinearAlgebra.AffineSpace.Independent
+import Mathlib.Tactic.Linarith
+
+/-!
+# Radon's convexity theorem
+
+Radon's theorem on convex sets states that any affine dependent set can be partitioned into two sets
+whose convex hulls intersect.
+
+## Main results
+
+* `Radon_partition`: Radon convexity theorem
+
+## Tags
+convex hull, radon
+
+-/
+
+
+open Set Finset
+open BigOperators
+
+universe u
+
+variable {𝕜 : Type*} {E : Type u} [LinearOrderedField 𝕜] [AddCommGroup E] [Module 𝕜 E]
+
+/-- Any family `f` of affine dependent vectors contains a set `I` with the property that
+convex hulls of `I` and `Iᶜ` intersect. -/
+theorem radon_partition {ι : Type*} {f : ι → E}
+    (h : ¬AffineIndependent 𝕜 f) : ∃ (I : Set ι),
+    (Set.Nonempty ((convexHull 𝕜 (f '' I)) ∩ (convexHull 𝕜 (f '' Iᶜ)))) := by
+  unfold AffineIndependent at h; push_neg at h
+  rcases h with ⟨s1, w, h_wsum, h_vsum, nonzero_w_index, h1, h2⟩
+  let I : Finset ι := s1.filter (fun i => w i ≥ 0)
+  let J : Finset ι := s1.filter (fun i => w i < 0)
+  use (I : Set ι)
+
+  let weights_sum_I := ∑ i in I, w i
+  let weights_sum_J := ∑ i in J, w i
+  have h3 : weights_sum_I + weights_sum_J = 0 := by
+    rw [←h_wsum]; simp
+    have h3_aux := Finset.sum_filter_add_sum_filter_not s1 (fun i => 0 ≤ w i) w
+    simp at h3_aux; exact h3_aux
+
+  have h4 : (weights_sum_I > 0) ∧ (weights_sum_J < 0) := by
+    rcases (lt_or_gt_of_ne h2) with h_neg | h_pos
+    have h_J_neg : weights_sum_J < 0 := by
+      rw [←neg_pos, ←Finset.sum_neg_distrib]
+      let neg_w : ι → 𝕜 := fun i => -w i
+      have h4_aux : neg_w nonzero_w_index ≤ (Finset.sum J neg_w) := by
+        apply Finset.single_le_sum
+        intros i hi
+        simp; rw [Finset.mem_filter] at hi; linarith
+        simp; constructor; exact h1; exact h_neg
+      simp at h4_aux; simp; linarith
+    constructor
+    linarith
+    exact h_J_neg
+
+    have h_I_pos : weights_sum_I > 0 := by
+      have h4_aux : weights_sum_I ≥ w nonzero_w_index := by
+        apply Finset.single_le_sum
+        intros i hi
+        rw [Finset.mem_filter] at hi; exact hi.right
+        simp; constructor; exact h1; exact LT.lt.le h_pos
+      linarith
+    constructor
+    exact h_I_pos
+    linarith
+  rcases h4 with ⟨h_I_pos, h_J_neg⟩
```

Comment:
A few things here:
- as above, prefer `0 < _` to `_ > 0`
- reasoning by cases is a perfectly fine strategy, but here you can take a shortcut: the lemma [Finset.exists_pos_of_sum_zero_of_exists_nonzero](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/BigOperators/Order.html#Finset.exists_pos_of_sum_zero_of_exists_nonzero) tells you that there exists at least a positive term (basically by doing the case disjunction for you), so you get `h_I_neg` directly
- proving that a sum of nonnegative elements is positive by exhibiting a positive term is so common that we have a lemma for it, [Finset.sum_pos'](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/BigOperators/Order.html#Finset.sum_pos')
- with my other suggestions below, you actually only need `h_I_pos`.
```suggestion
  have h_I_pos : 0 < weights_sum_I := by
    rcases exists_pos_of_sum_zero_of_exists_nonzero _ h_wsum ⟨nonzero_w_index, h1, h2⟩
      with ⟨pos_w_index, h1', h2'⟩
    exact sum_pos'
      (by simp only [mem_filter, and_imp, imp_self, implies_true])
      ⟨pos_w_index, by simp only [mem_filter, h1', h2'.le, and_self, h2']⟩
```
Here's what the proof of `h_J_neg` would look like from there:
```lean
  have h_J_neg : weights_sum_J < 0 := by
    linarith
```

Reply:
It's very cool! Can you teach me how you find exactly which lemmas to use in `simp only`?

Reply:
`simp? [...]` tells you which lemmas are used, both from the set of simp lemmas and the manually provided ones. To be fair, in these case you could keep the condensed form (`by simp` on line 55, `by simp [h1', h2'.le, h2']` on line 56). The rule is that all `simp` calls that do *not* close the goal should be replaced by `simp only` to be more predictable.